coverage/historical/github/historical_coverage_report.py

```diff
@@ -18,157 +18,253 @@
+        with open('report.json', 'w') as f:
+            json.dump(report, f, indent=4)
+        sys.stderr.flush()
         return
-    total_coverage /= workspace_considered
-    print(f"Workspaces analyzed: {workspace_considered}")
-    print(f"Total coverage across all workspaces: {total_coverage * 100:.2f}")
+
+    for workspace in report['workspaces']:
+        print(f"[Workspace: {workspace['name']}] coverage: {workspace['coverage'] * 100:.2f} across {workspace['repos_considered']} repositories")
+        for repo in workspace['repositories']:
+            print(f"  [Repo: {repo['name']}] coverage: {repo['coverage'] * 100:.2f} across {repo['prs_considered']} pull requests")
+    report['total_coverage'] /= report['workspace_considered']
+    print(f"Workspaces analyzed: {report['workspace_considered']}")
+    print(f"Total coverage across all workspaces: {report['total_coverage'] * 100:.2f}")
+    with open('report.json', 'w') as f:
+        json.dump(report, f, indent=4)
+    sys.stderr.flush()
         
-def process_workspace(total_coverage, workspace_considered, workspace, token):
+def process_workspace(report, workspace, token):
     """
     Process each workspace and calculate the coverage.
     """
-    print("="*30 + "\n" + f"Workspace: {workspace}")
+    keyboard_interrupted = False
     all_repositories = get_repositories(token, workspace)
     selected_repositories = prompt_for_repositories(all_repositories)
 
-    workspace_coverage = 0
-    repos_considered = 0
-    for repo in selected_repositories:
-        try:
-            workspace_coverage, repos_considered = process_repo(workspace_coverage, repos_considered, repo, workspace, token)
-        except Exception as e:
-            print(f"    [ERROR] Something went wrong while processing the repo {workspace}/{repo}:", e)
-            continue
-    if repos_considered == 0:
-        print(f"  No valid repos in workspace {workspace}, not including in coverage calculation")
-        return (total_coverage, workspace_considered)
-    workspace_considered += 1
-    workspace_coverage /= repos_considered
-    total_coverage += workspace_coverage
-    print(f"Total repos considered = {repos_considered}, total_coverage = {total_coverage}")
-    print(f"######################### Coverage percentage for Workspace {workspace} = {workspace_coverage * 100:.2f}% #####################################")
-    print("\n\n\n\n")
-    return (total_coverage, workspace_considered)
-
-def process_repo(workspace_coverage, repos_considered, repo, workspace, token):
+    workspace_report = {
+        'name': workspace,
+        'coverage': 0,
+        'repos_considered': 0,
+        'repositories': [],
+        'skipped_repos': []
+    }
+    try:
+        for repo in tqdm(selected_repositories, desc=f"  Processing repos in {workspace}:", leave=False, file=sys.stdout):
+            try:
+                process_repo(workspace_report, repo, workspace, token)
+            except Exception as e:
```

Comment:
please add exception handling for `keyboardInterrupt` as well `except KeyboardInterrupt as ki:
                raise KeyboardInterrupt(*ki.args)`